src/agent0/core/hyperdrive/interactive/i_hyperdrive.py

```diff
@@ -239,91 +250,65 @@ def _add_funds(
         # TODO expose async here to the caller eventually
-        trade_results: list[TradeResult] = asyncio.run(
-            async_execute_agent_trades(self.interface, agent, liquidate=False, interactive_mode=True)
-        )
-        tx_receipt = self._handle_trade_result(trade_results)
+        trade_result: TradeResult = asyncio.run(async_execute_single_trade(self.interface, agent, trade_object))
+        tx_receipt = self._handle_trade_result(trade_result)
         return self._build_event_obj_from_tx_receipt(HyperdriveActionType.ADD_LIQUIDITY, tx_receipt)
 
     def _remove_liquidity(self, agent: HyperdriveAgent, shares: FixedPoint) -> RemoveLiquidity:
-        # Set the next action to open a long
-        assert isinstance(agent.policy, IHyperdrivePolicy)
-        agent.policy.set_next_action(HyperdriveActionType.REMOVE_LIQUIDITY, shares)
+        trade_object = remove_liquidity_trade(shares)
         # TODO expose async here to the caller eventually
-        trade_results: list[TradeResult] = asyncio.run(
-            async_execute_agent_trades(self.interface, agent, liquidate=False, interactive_mode=True)
-        )
-        tx_receipt = self._handle_trade_result(trade_results)
+        trade_result: TradeResult = asyncio.run(async_execute_single_trade(self.interface, agent, trade_object))
+        tx_receipt = self._handle_trade_result(trade_result)
         return self._build_event_obj_from_tx_receipt(HyperdriveActionType.REMOVE_LIQUIDITY, tx_receipt)
 
     def _redeem_withdraw_share(self, agent: HyperdriveAgent, shares: FixedPoint) -> RedeemWithdrawalShares:
-        # Set the next action to open a long
-        assert isinstance(agent.policy, IHyperdrivePolicy)
-        agent.policy.set_next_action(HyperdriveActionType.REDEEM_WITHDRAW_SHARE, shares)
+        trade_object = redeem_withdraw_shares_trade(shares)
         # TODO expose async here to the caller eventually
-        trade_results: list[TradeResult] = asyncio.run(
-            async_execute_agent_trades(self.interface, agent, liquidate=False, interactive_mode=True)
-        )
+        trade_results: TradeResult = asyncio.run(async_execute_single_trade(self.interface, agent, trade_object))
         tx_receipt = self._handle_trade_result(trade_results)
         return self._build_event_obj_from_tx_receipt(HyperdriveActionType.REDEEM_WITHDRAW_SHARE, tx_receipt)
 
     def _execute_policy_action(
         self, agent: HyperdriveAgent
     ) -> list[OpenLong | OpenShort | CloseLong | CloseShort | AddLiquidity | RemoveLiquidity | RedeemWithdrawalShares]:
-        assert isinstance(agent.policy, IHyperdrivePolicy)
         # Only allow executing agent policies if a policy was passed in the constructor
-        if agent.policy.sub_policy is None:
+        # we check type instead of isinstance to explicitly check for the hyperdrive base class
+        # pylint: disable=unidiomatic-typecheck
+        if type(agent.policy) == HyperdriveBasePolicy:
```

Comment:
I think it would be a good idea to add a unit test for this to verify functionality